datafusion/physical-expr/src/intervals/cp_solver.rs

```
@@ -0,0 +1,897 @@
+                    child.interval.cast_to(cast_type, cast_options)?;
+            }
+        }
+        Ok(self.graph.index(self.root).interval.clone())
+    }
+
+    /// Updates/shrinks bounds for leaf expressions using interval arithmetic
+    /// via a top-down traversal.
+    fn propagate_constraints(
+        &mut self,
+        expr_stats: &mut [(usize, Interval)],
+    ) -> Result<PropagationResult> {
+        let mut bfs = Bfs::new(&self.graph, self.root);
+        while let Some(node) = bfs.next(&self.graph) {
+            // Get plan
+            let input = self.graph.index(node);
+            // Get calculated interval. BinaryExpr will propagate the interval according to
+            // this.
+            let node_interval = input.interval().clone();
+            let index = node.index();
+            if let Some((_, interval)) = expr_stats.iter_mut().find(|(e, _)| *e == index)
+            {
+                *interval = node_interval;
+                continue;
+            }
+
+            // Get outgoing edges (i.e. children):
+            let mut edges = self.graph.neighbors_directed(node, Outgoing).detach();
+
+            let expr_any = input.expr.as_any();
+            if let Some(binary) = expr_any.downcast_ref::<BinaryExpr>() {
+                // Get children intervals (note that we traverse in reverse):
+                let right_child_idx = edges.next_node(&self.graph).unwrap();
+                let right_interval = self.graph.index(right_child_idx).interval();
+                let left_child_idx = edges.next_node(&self.graph).unwrap();
+                let left_interval = self.graph.index(left_child_idx).interval();
+
+                let op = binary.op();
+                if let (Some(new_left_interval), Some(new_right_interval)) =
+                    if op.is_logic_operator() {
+                        // There is no propagation process for logical operators.
+                        continue;
```

I don't understand this comment. We can propagate over logical operators too, it mathematically works similarly. If the intent is to say that this is not implemented yet, you can return an error saying so.